README.md
## dotnet

This container will server an ASP.net application, via openresty
with JSON-logging and embedded WAF.

Place the app in /app.

There are 2 containers built:

- dotnet  --> dotnetcore 3.1 for modern applications (and .NET on Mono for legacy)
- dotnetw --> .NET on Mono on Wine for legacy applications that need win32

Each works the same way: [openresty](https://openresty.org/) and [lua-resty-waf](https://github.com/p0pr0ck5/lua-resty-waf)
run on port 5000

## Use (dotnetcore)

Create a Dockerfile for your application as:

```
FROM mcr.microsoft.com/dotnet/core/sdk:3.1 AS build-env
WORKDIR /app

COPY * /src
WORKDIR /src
RUN dotnet restore

# Copy everything else and build
COPY . /src
RUN dotnet publish -c release -o /app

FROM agilicus/dotnet
WORKDIR /app
COPY --from=build-env /app /app
```

and then build and test as:

```
tag=$(date +%Y%m%d_%H%M%S)
docker build -t MYNAME:$tag .
docker run --rm -it -p 4200:5000 MYNAME:$tag
```

and then open your browser to `http://localhost:4200`

## Use (.NET)

Create a Dockerfile for your application as:

```
FROM mono:5 AS build-env
WORKDIR /app

COPY ./ /app

RUN nuget restore -PackagesDirectory ../packages

RUN msbuild  /p:Configuration=Release\;WebOutputDir=/tmp/x\;OutDir=/tmp/x MYPROJECT.sln

FROM agilicus/dotnet
WORKDIR /app
COPY --from=build-env --chown=dotnet /tmp/x/_PublishedWebsites/MYAPP .
```

You may test it with:
```
tag=$(date +%Y%m%d_%H%M%S)
docker build -t MYNAME:$tag .
docker run --rm -it -p 4200:5000 MYNAME:$tag
```

and then open your browser to `http://localhost:5000`

## Environment variables

- HDR_CONTENT_SECURITY_POLICY
- HDR_FEATURE_POLICY

It will include configuration in server-scope from /rules/server.d/*.
It will include configuration in location-scope from /rules/location.d/*.


## The dotnetw wine prefix

The wine prefix is too complicated to install inside docker. It is built
outside, and a blob is placed in gcs.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.35](https://git.agilicus.com/open-source/dotnet/compare/v0.0.34...v0.0.35) (2020-09-01)


### Bug Fixes

* add ,--single-process,--no-zygote for zombie ([248a2a9](https://git.agilicus.com/open-source/dotnet/commit/248a2a9aaa790f8aa76140e06c52c8ab7e63572f))

### [0.0.34](https://git.agilicus.com/open-source/dotnet/compare/v0.0.33...v0.0.34) (2020-08-31)


### Features

* add jsreports library support for chrome ([9152565](https://git.agilicus.com/open-source/dotnet/commit/9152565d6f25d234d25a9d5b6d38be3554686114))

### [0.0.33](https://git.agilicus.com/open-source/dotnet/compare/v0.0.32...v0.0.33) (2020-08-26)


### Bug Fixes

* add libgdiplus for rendering ([21aa68d](https://git.agilicus.com/open-source/dotnet/commit/21aa68d1d20e154272b741cb49bfa4d86e3ef2b5))

### [0.0.32](https://git.agilicus.com/open-source/dotnet/compare/v0.0.31...v0.0.32) (2020-08-19)


### Bug Fixes

* allow 404 to bring online ([263872f](https://git.agilicus.com/open-source/dotnet/commit/263872fb14b014db9efd9203ed3baa306f56fd2e))

### [0.0.31](https://git.agilicus.com/open-source/dotnet/compare/v0.0.30...v0.0.31) (2020-08-19)


### Bug Fixes

* must use userid 1000 ([0f5acac](https://git.agilicus.com/open-source/dotnet/commit/0f5acac41ef45451e852dfdd330877295026531a))

### [0.0.30](https://git.agilicus.com/open-source/dotnet/compare/v0.0.29...v0.0.30) (2020-08-14)


### Bug Fixes

* add dotnetcore 2.2, Index.html to path ([6ab4ee4](https://git.agilicus.com/open-source/dotnet/commit/6ab4ee4f08d1f9d98f0ab03b4d6830483f644fbe))

### [0.0.29](https://git.agilicus.com/open-source/dotnet/compare/v0.0.28...v0.0.29) (2020-08-14)


### Features

* add override DOTNET/DOTNETCORE env var ([8e2b057](https://git.agilicus.com/open-source/dotnet/commit/8e2b057897f479dd6a174e0bfdaf3393715965a8))
* add override DOTNET/DOTNETCORE env var ([91875d2](https://git.agilicus.com/open-source/dotnet/commit/91875d277bb0bd007b55308c898736c76c0e8058))

### [0.0.28](https://git.agilicus.com/open-source/dotnet/compare/v0.0.27...v0.0.28) (2020-08-13)


### Bug Fixes

* correct shared dotnet/.net entry ([7246b03](https://git.agilicus.com/open-source/dotnet/commit/7246b032ff9a44487ec9a2879f5df9fa6953a89e))

### [0.0.27](https://git.agilicus.com/open-source/dotnet/compare/v0.0.26...v0.0.27) (2020-08-13)


### Features

* merge dotnetcore and mono containers ([3e7d11e](https://git.agilicus.com/open-source/dotnet/commit/3e7d11e0d291f11108b66896f12ccc9d7ef1f22a))

### [0.0.26](https://git.agilicus.com/open-source/dotnet/compare/v0.0.25...v0.0.26) (2020-08-11)


### Bug Fixes

* refactor to use openresty default nginx.conf ([7a1288d](https://git.agilicus.com/open-source/dotnet/commit/7a1288dd7255aa441bd85c556cfb3b79cd5c4031))

### [0.0.25](https://git.agilicus.com/open-source/dotnet/compare/v0.0.24...v0.0.25) (2020-07-22)


### Bug Fixes

* add sub field from jwt to log ([16f5223](https://git.agilicus.com/open-source/dotnet/commit/16f5223a7d60d26b1cf86774ca8b985c8afd7f5b))

### [0.0.24](https://git.agilicus.com/open-source/dotnet/compare/v0.0.23...v0.0.24) (2020-06-27)


### Bug Fixes

* add client_max_body_size=32M ([6d8dd98](https://git.agilicus.com/open-source/dotnet/commit/6d8dd98280031a6226e0f08a7017171db5dc4fec))

### [0.0.23](https://git.agilicus.com/open-source/dotnet/compare/v0.0.22...v0.0.23) (2020-06-22)


### Bug Fixes

* add seatbelts and checks around entry detection for dotnetcore ([405a300](https://git.agilicus.com/open-source/dotnet/commit/405a30027f446ba85897ed7355b0e64b5b061c74))

### [0.0.22](https://git.agilicus.com/open-source/dotnet/compare/v0.0.21...v0.0.22) (2020-04-29)


### Bug Fixes

* enable parallel build in CI ([91eb8f9](https://git.agilicus.com/open-source/dotnet/commit/91eb8f9479f9ff9ec4d03f5df17d24f3e68404e2))

### [0.0.21](https://git.agilicus.com/open-source/dotnet/compare/v0.0.20...v0.0.21) (2020-04-29)


### Bug Fixes

* add real remote_ip and x_request_id to log ([2f0b4c6](https://git.agilicus.com/open-source/dotnet/commit/2f0b4c64474f1253938881f5be0bd6882c1bc26c))

### [0.0.20](https://git.agilicus.com/open-source/dotnet/compare/v0.0.19...v0.0.20) (2020-04-28)


### Bug Fixes

* allow override entry point for hybrid dotnetcore + .net ([93ceec8](https://git.agilicus.com/open-source/dotnet/commit/93ceec8cdc88ea928cae9307c6f615182a1d98c4))
* redirect entrypoint to /dev/null except for stderr ([047a35e](https://git.agilicus.com/open-source/dotnet/commit/047a35e122b8013d4fbf862688dc5228840e02b4))

### [0.0.19](https://git.agilicus.com/open-source/dotnet/compare/v0.0.18...v0.0.19) (2020-04-28)


### Bug Fixes

* add dotnet 4.7.2 ([4a89461](https://git.agilicus.com/open-source/dotnet/commit/4a89461b1e06990a72ea50ad842dba1c7bd4b200))

### [0.0.18](https://git.agilicus.com/open-source/dotnet/compare/v0.0.17...v0.0.18) (2020-04-24)


### Features

* add .net 4.8 support ([1e424eb](https://git.agilicus.com/open-source/dotnet/commit/1e424eb7d85ae0c5127a405c3ad690359556db48))

### [0.0.17](https://git.agilicus.com/open-source/dotnet/compare/v0.0.16...v0.0.17) (2020-03-27)


### Bug Fixes

* add 302 as valid start condition ([1355720](https://git.agilicus.com/open-source/dotnet/commit/135572065f71ed29da2623dedf08239077bd7a1a))

### [0.0.16](https://git.agilicus.com/open-source/dotnet/compare/v0.0.15...v0.0.16) (2020-03-03)


### Bug Fixes

* correct dotnetcore entrypoint ([f6f70cc](https://git.agilicus.com/open-source/dotnet/commit/f6f70cc76de5f445b9378ea02808768bfc91ef79))

### [0.0.15](https://git.agilicus.com/open-source/dotnet/compare/v0.0.14...v0.0.15) (2020-02-25)


### Bug Fixes

* harmonise logging w/ istio proxy ([ea7d15e](https://git.agilicus.com/open-source/dotnet/commit/ea7d15e74f484ac26376f28e2d6eacfa9690fcb7))

### [0.0.14](https://git.agilicus.com/open-source/dotnet/compare/v0.0.13...v0.0.14) (2020-02-22)



entry.sh
#!/usr/bin/dumb-init /bin/bash
log() {
  >&2 echo "{ \"time\": \"$(date -Is)\", \"log\": \"$*\"}"
}

wait_ready() {
    set -a
    SCRIPT_NAME=/
    SCRIPT_FILENAME=/
    QUERY_STRING= REQUEST_URI=/
    REQUEST_METHOD=GET
    until cgi-fcgi -bind -connect 127.0.0.1:9000 |egrep -q "Status: "
    do
        log "Wait for fastcgi-mono to be ready"
        sleep 1
    done
}

wait_ready_dnc() {
    set -a
    ntries=0
    SCRIPT_NAME=/
    SCRIPT_FILENAME=/
    QUERY_STRING= REQUEST_URI=/
    REQUEST_METHOD=GET
    until curl -sq localhost:9000 >/dev/null 2>&1
    do
        ntries=$((ntries+1))
        if [ $ntries -gt 30 ]
        then
	    log "Error: dotnet application didn't start (check ENTRY environment variable)"
            exit 1
        fi
        log "Wait for dotnetcore to be ready"
        sleep 1
    done
}

export ASPNETCORE_URLS=http://0.0.0.0:9000

is_dotnetcore() {
    if [ -n "$DOTNET" ]
    then
      log "DOTNET is is set, force .NET xsp runtime"
      return 1
    fi
    if [ -n "$DOTNETCORE" ]
    then
      log "DOTNETCORE is is set, force dotnetcore runtime"
      return 0
    fi

    if [ -f web.config -o -f Web.config ]
    then
	grep -q aspNetCore [wW]eb.config
        res=$?
        if [ $res -eq 0 ]
        then
	    log "[wW]eb.config exists, is dotnetcore"
            return 0
        fi
        log "[wW]eb.config exists, is not dotnetcore"
        return 1
    fi
    # We should have 1 FOO.runtimeconfig.json
    n_rtc=$(ls *runtimeconfig.json 2>/dev/null | wc -l)
    if [ $n_rtc -eq 0 ]
    then
        log "This is not dotnetcore: no FOO.runtimeconfig.json"
	return 1
    fi
    if [ -n "$ENTRY" -o $n_rtc -eq 1 ]
    then
        log "dotnetcore: ENTRY=\"$ENTRY\" set or runtimeconfig.json exists"
        return 0
    fi
    log "Fall through: assume not dotnetcore"
    return 1
}

do_dotnet() {
    log "Assume .NET runtime"
    log "Start fastcgi"
    mv /rules/server.d/root.conf.xsp /rules/server.d/root.conf
    fastcgi-mono-server4 --applications=/:. --socket=tcp:127.0.0.1:9000 &
    wait_ready
}

do_dotnetcore() {
    log "Assume dotnetcore runtime"
    mv /rules/server.d/root.conf.dnc /rules/server.d/root.conf
    if [ -z "$ENTRY" ]
    then
        n_rtc=$(ls *runtimeconfig.json 2>/dev/null | wc -l)
        if [ $n_rtc -eq 1 ]
        then
	    rtc=$(ls *.runtimeconfig.json)
            ENTRY=${rtc%.runtimeconfig.json}.dll
            log "dotnetcore: chose ENTRY=\"$ENTRY\" via runtimeconfig.json method"
        else
 
	    pp=$(xml2 < [wW]eb.config | awk -F =  '/processPath/ { sub("^[^a-zA-Z0-9]*", "", $2); print $2}')
	    pa=$(xml2 < [wW]eb.config | awk -F =  '/arguments/ { sub("^[^a-zA-Z0-9]*", "", $2); print $2}')
	    log "processPath: <<$pp>>, arguments: <<$pa>>"
	    if [ "$pp" = "dotnet" ]
	    then
		ENTRY="$pa"
	    else
		ENTRY="$pp"
	    fi
        fi
    fi
    if [ ! -f "$ENTRY" ]
    then
	log "Error: <<$ENTRY>> does not exist in /app directory. In /app these files exist:"
	ls
	exit 1
    fi
    log "Start dotnetcore for <<$ENTRY>>"
    dotnet $ENTRY &
}

is_dotnetcore

if [ $? -eq 0 ]
then
    log "dotnetcore: is_dotnetcore return 0"
    do_dotnetcore
else
    log "not dotnetcore: is_dotnetcore return != 0"
    do_dotnet
fi

log "Start openresty"
openresty -g 'daemon off;' &
wait -n
log "Exit..."


entry-dotnetcore.sh
#!/usr/bin/dumb-init /bin/bash
log() {
  >&2 echo "{ \"time\": \"$(date -Is)\", \"log\": \"$*\"}"
}

wait_ready() {
    set -a
    ntries=0
    SCRIPT_NAME=/
    SCRIPT_FILENAME=/
    QUERY_STRING= REQUEST_URI=/
    REQUEST_METHOD=GET
    until curl -sq localhost:9000 >/dev/null 2>&1
    do
        ntries=$((ntries+1))
        if [ $ntries -gt 30 ]
        then
	    log "Error: dotnet application didn't start (check ENTRY environment variable)"
            exit 1
        fi
        log "Wait for dotnetcore to be ready"
        sleep 1
    done
}

export ASPNETCORE_URLS=http://0.0.0.0:9000

# Make an assumption the most recent dll is the entrypoint
#ENTRY=$(ls -tr *.runtimeconfig.json | tail -1)
#ENTRY=${ENTRY%.runtimeconfig.json}.dll
if [ -z "$ENTRY" ]
then
    pp=$(xml2 < [wW]eb.config | awk -F =  '/processPath/ { sub("^[^a-zA-Z0-9]*", "", $2); print $2}')
    pa=$(xml2 < [wW]eb.config | awk -F =  '/arguments/ { sub("^[^a-zA-Z0-9]*", "", $2); print $2}')
    log "processPath: <<$pp>>, arguments: <<$pa>>"
    if [ "$pp" = "dotnet" ]
    then
        ENTRY="$pa"
    else
        ENTRY="$pp"
    fi
fi
if [ ! -f "$ENTRY" ]
then
    log "Error: <<$ENTRY>> does not exist in /app directory"
    log "In /app, I see:"
    ls
    exit 1
fi
log "Start dotnetcore for <<$ENTRY>>"
dotnet $ENTRY &
wait_ready
log "Start openresty"
openresty -g 'daemon off;' &
wait -n
log "Exit..."



entry-dotnetw.sh
#!/usr/bin/dumb-init /bin/bash
log() {
  >&2 echo "{ \"time\": \"$(date -Is)\", \"log\": \"$*\"}"
}
wait_ready() {
    set -a
    SCRIPT_NAME=/
    SCRIPT_FILENAME=/
    QUERY_STRING= REQUEST_URI=/
    REQUEST_METHOD=GET
    until cgi-fcgi -bind -connect 127.0.0.1:9000 |egrep -q "Status: 200 OK|Status: 404 Not Found|Status: 302 Found"
    do
        log "Wait for fastcgi-mono to be ready"
        sleep 1
    done
}

export ASPNETCORE_URLS=http://0.0.0.0:9000

log "Startup"
#rm -f /tmp/.X1-lock
#log "Start Xvfb"
#Xvfb :1 &
#log "Update .NET bits (foreground) ..."
#wine /home/dotnet/.wine/drive_c/windows/Microsoft.NET/Framework/v4.0.30319/ngen.exe update /nologo /force ||true
log "Start fastcgi"
if [ -n "$ENTRY" ]
then
    # A hybrid .net + dotnetcore, created as a Kestrel executable.
    cp /rules/server.d/root.dnc /rules/server.d/root.conf
    wine $ENTRY >/dev/null &
else
    wine /home/dotnet/.wine/drive_c/windows/Microsoft.NET/assembly/GAC_32/fastcgi-mono-server4/v4.0_0.4.0.0__0738eb9f132ed756/fastcgi-mono-server4.exe --applications=/:. --socket=tcp:127.0.0.1:9000 &
    wait_ready
fi
log "Start openresty"
openresty -g 'daemon off;' &
wait -n
log "Exit..."
